File: splatter.py
from random import weibullvariate
import numpy as np

def splat_conv2d(input, kernel, mode=''):
    #this function skips reversing the kernel
    #establish size of input image
    # print("starting convolution")
    rows = input.shape[0]
    cols = input.shape[1]

    #establish weight kernel size
    weightSize = kernel.shape[0]

    #establish weight indexing
    odd = False
    weightIndex = int(weightSize/2) 
    if(weightSize%2):
        odd = True
    #pad input matrix
    padInput = np.zeros((rows + weightIndex*2, cols + weightIndex*2))
    padInput[weightIndex:-weightIndex, weightIndex:-weightIndex] = input

    #create padded output image
    output = np.zeros_like(padInput)


    #iterate though input matrix
    for i in range(weightIndex, weightIndex+rows): #iterate through rows of input
        for j in range(weightIndex, weightIndex+cols): #iterate through columns of input
            
            #check if index is nonzero
            if(padInput[i][j] > 0 or padInput[i][j] <0):
                #update output using vector operations
                if(odd):
                    output[i-weightIndex:i+weightIndex +1,j-weightIndex:j+weightIndex+1] += padInput[i][j]*kernel
                else:
                     output[i-weightIndex:i+weightIndex,j-weightIndex:j+weightIndex] += padInput[i][j]*kernel


    #adjust final output size using splicing
    if(mode == "full"):
        if(odd):
            return output
        else:
            return output[:-1, :-1]
    elif(mode == "valid"):
        if(cols<weightSize or rows< weightSize):
            if(odd): 
                return output[rows-1:-rows+1, cols-1:-cols+1]
            else:
                return output[rows-1:-rows, cols-1:-cols]
        if(odd):
            return output[2*weightIndex:-2*weightIndex, 2*weightIndex:-2*weightIndex]
        else:
            return output[2*weightIndex-1:-2*weightIndex ,2*weightIndex-1:-2*weightIndex]
    output = output[weightIndex:-weightIndex, weightIndex:-weightIndex]
    
    return output
    #end of function

def splat_corr2d(input, kernel, mode=''):
   

   
    # print("starting convolution")
    kernel = np.flipud(np.fliplr(kernel))

    rows = input.shape[0]
    cols = input.shape[1]

    #establish weight kernel size
    weightSize = kernel.shape[0]

    odd = False
    if(weightSize%2):
        odd = True

    #establish weight indexing
    weightIndex = int(weightSize/2) 

    #pad input matrix
    padInput = np.zeros((rows + weightIndex*2, cols + weightIndex*2))
    padInput[weightIndex:-weightIndex, weightIndex:-weightIndex] = input

    #create padded output image
    output = np.zeros_like(padInput)


    #iterate though input matrix
    for i in range(weightIndex, weightIndex+rows): #iterate through rows of input
        for j in range(weightIndex, weightIndex+cols): #iterate through columns of input
            
            #check if index is nonzero
            if(padInput[i][j] > 0 or padInput[i][j] <0):

                #update output using vector operation
                
                if(odd):
                    output[i-weightIndex:i+weightIndex +1,j-weightIndex:j+weightIndex+1] += padInput[i][j]*kernel
                else:
                    output[i-weightIndex:i+weightIndex ,j-weightIndex:j+weightIndex] += padInput[i][j]*kernel


    #adjust final output size using splicing
    if(mode == "full"):
        return output
    elif(mode == "valid"):
        if(odd):
            return output[2*weightIndex:-2*weightIndex, 2*weightIndex:-2*weightIndex]
        else:
            return output[2*weightIndex-1:-2*weightIndex ,2*weightIndex-1:-2*weightIndex]
    output = output[weightIndex:-weightIndex, weightIndex:-weightIndex]
    
    return output
    #end of function

File: splatter.c
#include <stdlib.h>
#include <stdio.h>


//correlation forward
void splatterForward(int rows, int cols, int kernelSize,double **input, double **output, double **kernel ){
/*
currently only works for odd kernels
*/
    
    int weightIndex = kernelSize -2;

   

    //rows
    for(int i = weightIndex; i < weightIndex+rows; i++){
        //cols
        for(int j = weightIndex; j < weightIndex+cols; j++){
            
            //check if nonzero
            // float temp = input[i][j];
            if(input[i][j] != 0){
                
                for(int k = -1*weightIndex; k <= weightIndex; k++){
                    for(int l = -1*weightIndex; l <= weightIndex; l++){
                        
                        // printf("%d %d %d %d\n", i, j, k, l);
                        // float kern = kernel[k][l];
                        output[i-k][j-l] += kernel[1+k][1+l] * input[i][j];
                        
                        
                        // printf("%d %d %f\n", i, j, input[i][j]);
                        // printf("%d %d %f\n", k, l, kernel[1+k][1+l]);
                        // printf("%f\n\n", output[i][j]);

                        
                    }
                }
            }
        }
    }

}




//correlation forward
void splatterBackwardInput(int rows, int cols, int kernelSize,double **input, double **output, double **kernel ){
/*
currently only works for odd kernels
*/
    
    int weightIndex = kernelSize/2;

   

    //rows
    for(int i = weightIndex; i < weightIndex+rows; i++){
        //cols
        for(int j = weightIndex; j < weightIndex+cols; j++){
            
            //check if nonzero
            // float temp = input[i][j];
            if(input[i][j] != 0){
                
                for(int k = -1*weightIndex; k <= weightIndex; k++){
                    for(int l = -1*weightIndex; l <= weightIndex; l++){
                        
                        // printf("%d %d %d %d\n", i, j, k, l);
                        // float kern = kernel[k][l];
                        output[i+k][j+l] += kernel[weightIndex+k][weightIndex+l] * input[i][j];
                        
                        
                        // printf("%d %d %f\n", i, j, input[i][j]);
                        // printf("%d %d %f\n", k, l, kernel[weightIndex+k][weightIndex+l]);
                        // printf("%f\n\n", output[i][j]);

                        
                    }
                }
            }
        }
    }

}



//correlation forward
void splatterBackwardFilter(int rows, int cols, int kernelSize,double **input, double **output, double **kernel ){
/*
currently only works for odd kernels
*/
    
    int weightIndex = kernelSize/2;

   

    //rows
    for(int i = weightIndex; i < weightIndex+rows; i++){
        //cols
        for(int j = weightIndex; j < weightIndex+cols; j++){
            
            //check if nonzero
            // float temp = input[i][j];
            
            if(input[i][j] != 0){
                
                for(int k = -1*weightIndex; k < weightIndex; k++){
                    for(int l = -1*weightIndex; l < weightIndex; l++){
                        
                        // printf("%d %d %d %d\n", i, j, k, l);
                        // float kern = kernel[k][l];
                        output[i-k][j-l] += kernel[weightIndex+k][weightIndex+l] * input[i][j];
                        
                        
                        // printf("%d %d %f\n", i, j, input[i][j]);
                        // printf("%d %d %f\n", k, l, kernel[weightIndex+k][weightIndex+l]);
                        // printf("%f\n\n", output[i][j]);

                        
                    }
                }
            }
        }
    }

}